### ISSUE TITLE
False positive for [E0308] (with StreamExt::filter_map)

### ISSUE THREAD

**Issue by aldanor** [intellij-rust/intellij-rust#6362]
## Environment

* **IntelliJ Rust plugin version:** 0.3.131.3366-201
* **Rust toolchain version:** 1.46.0
* **IDE name and version:** CLion 2020.1.1
* **Operating system:** macOS 10.14.6

## Problem description

Example below compiles fine, but is marked as E0308 ([link to `StreamExt::filter_map()`](https://docs.rs/futures/0.3.7/futures/stream/trait.StreamExt.html#method.filter_map)). There might be a simpler example, but that's what I managed to reduce it to so far.

## Steps to reproduce

```rust
use futures::{stream::{Stream, StreamExt}, future::ready};

struct A;
struct B;

fn f(_: A) -> Option<B> {
    unimplemented!()
}

fn g(stream: impl Stream<Item = A>) -> impl Stream<Item = B> {
    stream.filter_map(|x| ready(f(x)))
//                        ~~~~~~~~~~~ expected Option<<unknown>>, found Ready<Option<B>>
}
```

**Comment by aldanor** [comment]
Upd: looks like it's confused with `.filter_map()` belonging to `StreamExt`, since this seems to work:

```rust
StreamExt::filter_map(stream, |x| ready(f(x)))
```

**Comment by mati865** [comment]
Another example that might be useful:
```rust
fn foo(a: &[usize]) -> impl Stream<Item = &usize> + '_ {
    stream::iter(a)
}

fn bar(a: &[usize]) -> impl Stream<Item = (&usize, &usize)> + '_ {
    foo(a).map(|b| (b, b)) // error: expected `Result<<unknown>, <unknown>>`, found `(<unknown>, <unknown>)` 
}

fn foobar(a: &[usize]) -> impl Stream<Item = (&usize, &usize)> + '_ {
    stream::iter(a).map(|b| (b, b)) // inlined `foo` works fine
}
```
Notice how `b` cannot be inferred in `bar` but in `foobar` it's correctly hinted as `&usize`:
![obraz](https://user-images.githubusercontent.com/1174646/205684376-0ed68603-6418-4e7d-a799-c383bd4d6d49.png)

<details>
<summary>Environment</summary>

* **IntelliJ Rust plugin version:** 0.4.184.5057-223
* **Rust toolchain version:** 1.67.0-nightly (53e4b9dd7 2022-12-04) x86_64-unknown-linux-gnu
* **IDE name and version:** CLion 2022.3 (CL-223.7571.171)
* **Macro expansion:** enabled
* **Additional experimental features:** org.rust.macros.proc.attr, org.rust.cargo.features.settings.gutter, org.rust.cargo.emulate.terminal, org.rust.macros.proc

</details>

**Comment by Maksim20023** [comment]
Hi! I’m closing this ticket because the issue does not reproduce in the current environment (CLion 2023.2 Build #CL-232.8660.186, Rust plugin 0.4.199.5415-232). If this issue still reproduces for you, please reopen the ticket.
